marimo/_runtime/copy.py

```diff
@@ -0,0 +1,174 @@
+# Copyright 2024 Marimo. All rights reserved.
+from __future__ import annotations
+
+import inspect
+import sys
+import weakref
+from copy import copy
+from typing import (
+    Any,
+    Callable,
+    Generic,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+
+T = TypeVar("T")
+if sys.version_info < (3, 9):
+    # Future does not seem to work for this in CI.
+    Ref = Union["weakref.ReferenceType[T]", Callable[[], T]]
+else:
+    Ref = Union[weakref.ReferenceType[T], Callable[[], T]]
+
+
+class CloneError(Exception):
+    """Thrown when strict execution fail to deep copy or clone."""
+
+
+class RoError(Exception):
+    """Thrown when attempting to modify a read-only object."""
+
+
+class _Copy(Generic[T]):
+    """Base wrapper class for strict execution."""
+
+    __ref__: Ref[T]
+
+
+class ZeroCopy(_Copy[T]):
+    """Wrapper class for strict execution (stops deepcopy)."""
+
+
+class ShallowCopy(_Copy[T]):
+    """Wrapper class for strict execution (does copy over deepcopy)."""
```

Comment:
I suppose this would go in the docs you'll write in a future PR (no need right now). But out of curiosity, when would someone choose `shallow_copy` over `zero_copy`?

Reply:
A container of object that's shouldnt be copied, like `[state, state, state]`

Reply:
> A container of object that's shouldnt be copied, like [state, state, state]

If you don't want something to be copied, wouldn't you use `zero_copy`, not `shallow_copy`?

Reply:
The container can mutate w/ ZeroCopy but not shallow copy

`states.append(state)`

will be reset w/ shallow copy